src/goto-symex/symex_clean_expr.cpp

```diff
@@ -94,41 +30,28 @@ void goto_symext::process_array_expr(exprt &expr)
   {
     if_exprt &if_expr=to_if_expr(expr);
     process_array_expr(if_expr.true_case());
+    process_array_expr(if_expr.false_case());
 
-    process_array_expr_rec(if_expr.false_case(),
-                           if_expr.true_case().type());
+    if(!base_type_eq(if_expr.true_case(), if_expr.false_case(), ns))
+    {
+      byte_extract_exprt be(
+        byte_extract_id(),
+        if_expr.false_case(),
+        from_integer(0, index_type()),
+        if_expr.true_case().type());
+
+      if_expr.false_case().swap(be);
+    }
 
     if_expr.type()=if_expr.true_case().type();
   }
-  else if(expr.id()==ID_index)
-  {
-    // strip index
-    index_exprt &index_expr=to_index_expr(expr);
-    exprt tmp=index_expr.array();
-    expr.swap(tmp);
-  }
-  else if(expr.id()==ID_typecast)
-  {
-    // strip
-    exprt tmp=to_typecast_expr(expr).op0();
-    expr.swap(tmp);
-    process_array_expr(expr);
-  }
   else if(expr.id()==ID_address_of)
   {
     // strip
     exprt tmp=to_address_of_expr(expr).op0();
     expr.swap(tmp);
     process_array_expr(expr);
   }
-  else if(expr.id()==ID_byte_extract_little_endian ||
-          expr.id()==ID_byte_extract_big_endian)
-  {
-    // pick the root object
-    exprt tmp=to_byte_extract_expr(expr).op();
-    expr.swap(tmp);
-    process_array_expr(expr);
-  }
   else if(expr.id()==ID_symbol &&
```

Comment:
There seem to be quite a few cases that have been refactored here that, to my eyes, are not covered by the test case added.